### PR TITLE
Add alias/merge methods to YAML::Builder and YAML::Nodes::Builder

### LINKED ISSUE #7821: YAML Anchors added before key
https://play.crystal-lang.org/#/r/6zj8

```crystal
require "yaml"

struct Config
  include YAML::Serializable

  def initialize; end

  getter name : String = "Default"
end

struct Environments
  include YAML::Serializable

  def initialize; end

  getter environments : Hash(String, Config) = {"development" => Config.new}

  def to_yaml(builder : YAML::Nodes::Builder)
    builder.mapping do
      builder.scalar "environments"
      builder.mapping do
        builder.scalar "development", reference: "development", anchor: "development"
        @environments["development"].to_yaml builder
      end
      
      builder.scalar "test"
      builder.mapping(anchor: "test") do
        builder.scalar "<<"
        builder.scalar "", reference: "development"
      end

      builder.scalar "production"
      builder.mapping(anchor: "production") do
        builder.scalar "<<"
        builder.scalar "", reference: "development"
      end
    end
  end
end

Environments.new.to_yaml
#---
#environments:
#  &development development:
#    name: Default
#test: &test
#  <<: *development
#production: &production
#  <<: *development

yaml = <<-YAML
---
environments:
  &development development:
    name: Foo
  test: &test
    <<: *development
  production: &production
    <<: *development
    name: Bar
YAML

dev = Environments.from_yaml(yaml).environments["development"]
test = Environments.from_yaml(yaml).environments["test"]
prod = Environments.from_yaml(yaml).environments["production"]

dev.name # => Foo
test.name # => Default
prod.name # => Bar
```

When building the structure, adding an anchor to a scalar value adds the anchor _before_ the key which causes the values to not be inherited correctly.

Moving the `&development` to _after_ the `development` key makes the test environment to also return `Foo`, which is the correct behavior.

I also tried manually defining the `to_yaml` within the `Config` class (in the my actual example) to add the anchor to the mapping.  In this case the anchor is added correctly, but seems to break the references.

### PR DESCRIPTION
**Pull request by Blacksmoke16**
Resolves #7821.

* Adds `alias` and `merge` methods to `YAML::Nodes::Builder`
* Adds `merge` method to `YAML::Builder`
* Also adds documentation blocks and specs for each

### CODE REVIEW

File: src/yaml/builder.cr [hunk 1]
```diff
@@ -130,11 +130,43 @@ class YAML::Builder
     yield.tap { end_mapping }
   end
 
-  def alias(anchor : String)
+  # Aliases *anchor*.
```
Comment (asterite) [inline src/yaml/builder.cr]: Maybe "Emits an alias to the given \*anchor\*"? Same in the other method.

Comment (Blacksmoke16) [reply src/yaml/builder.cr]: @asterite What about `Emits the extend key, `<<`, and an alias to the given *anchor*.` for `#extend`?

Comment (asterite) [reply src/yaml/builder.cr]: In the YAML spec it's called "merge": https://yaml.org/type/merge.html

I would probably rename `extend` to `merge`, say that it emits the scalar "<<" followed by an alias to the given anchor, and include a link to the spec.

Comment (Blacksmoke16) [reply src/yaml/builder.cr]: 👍 I could go with that.

File: spec/std/yaml/builder_spec.cr [hunk 2]
```diff
@@ -24,6 +24,12 @@ describe YAML::Builder do
     end
   end
 
+  it "writes alias" do
+    assert_built("--- *key\n") do
+      itself.alias "key"
```
Comment (Sija) [inline spec/std/yaml/builder_spec.cr]:
```suggestion
     alias "key"
```

File: spec/std/yaml/builder_spec.cr [hunk 3]
```diff
@@ -115,6 +121,23 @@ describe YAML::Builder do
     end
   end
 
+  it "writes mapping with alias" do
+    assert_built("---\nfoo: *bar\n") do
+      mapping do
+        scalar "foo"
+        itself.alias "bar"
```
Comment (Sija) [inline spec/std/yaml/builder_spec.cr]:
```suggestion
       alias "bar"
```

File: spec/std/yaml/nodes/builder_spec.cr [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+  end
+  string.should eq(expected)
+end
+
+describe YAML::Nodes::Builder do
+  describe "#alias" do
+    describe "as a scalar value" do
+      it "writes correctly" do
+        assert_built("--- *key\n") do
+          itself.alias "key"
```
Comment (Sija) [inline spec/std/yaml/nodes/builder_spec.cr]:
```suggestion
         alias "key"
```

File: spec/std/yaml/nodes/builder_spec.cr [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+        end
+      end
+    end
+
+    describe "within a mapping" do
+      it "writes correctly" do
+        assert_built("---\nfoo: *bar\n") do
+          mapping do
+            scalar "foo"
+            itself.alias "bar"
```
Comment (Sija) [inline spec/std/yaml/nodes/builder_spec.cr]:
```suggestion
           alias "bar"
```

File: src/yaml/builder.cr [hunk 6]
```diff
@@ -130,11 +130,45 @@ class YAML::Builder
...
+  # yaml = YAML.build do |builder|
+  #   builder.mapping do
+  #     builder.merge "development"
+  #   end
+  # end
+  #
+  # yaml # => "---\n<<: *development\n"
+  # ```
+  def merge(anchor : String) : Nil
+    self.scalar "<<"
```
Comment (Sija) [inline src/yaml/builder.cr]:
```suggestion
   scalar "<<"
```

File: src/yaml/builder.cr [hunk 7]
```diff
@@ -130,11 +130,45 @@ class YAML::Builder
...
+    self.alias anchor
```
Comment (Sija) [inline src/yaml/builder.cr]:
```suggestion
   alias anchor
```

File: src/yaml/nodes/builder.cr [hunk 8]
```diff
@@ -22,6 +22,52 @@ class YAML::Nodes::Builder
...
+  # end
+  #
+  # yaml = YAML.build do |builder|
+  #   nodes_builder.document.to_yaml builder
+  # end
+  #
+  # yaml # => "---\n<<: *key\n"
+  # ```
+  def merge(anchor : String) : Nil
+    self.scalar "<<"
```
Comment (Sija) [inline src/yaml/nodes/builder.cr]:
```suggestion
   scalar "<<"
```

File: src/yaml/nodes/builder.cr [hunk 9]
```diff
@@ -22,6 +22,52 @@ class YAML::Nodes::Builder
...
+    self.alias anchor
```
Comment (Sija) [inline src/yaml/nodes/builder.cr]:
```suggestion
   alias anchor
```

File: src/yaml/nodes/builder.cr [hunk 10]
```diff
@@ -22,6 +22,52 @@ class YAML::Nodes::Builder
     @anchor_count = 0
   end
 
+  # Emits an alias to the given *anchor*.
+  #
+  # ```crystal
```
Comment (Sija) [inline src/yaml/nodes/builder.cr]:
```suggestion
  # ```
```

File: src/yaml/builder.cr [hunk 11]
```diff
@@ -130,11 +130,45 @@ class YAML::Builder
...
+  def alias(anchor : String) : Nil
     LibYAML.yaml_alias_event_initialize(pointerof(@event), anchor)
     yaml_emit("alias")
   end
 
+  # Emits the scalar "<<" followed by an alias to the given *anchor*.
+  #
+  # See [YAML Merge](https://yaml.org/type/merge.html).
+  #
+  # ```crystal
```
Comment (Sija) [inline src/yaml/builder.cr]:
```suggestion
  # ```
```

File: src/yaml/nodes/builder.cr (hunk 10 above)
Comment (Sija) [inline src/yaml/nodes/builder.cr]:
```suggestion
  # ```
```

File: src/yaml/nodes/builder.cr [hunk 12]
```diff
@@ -22,6 +22,52 @@ class YAML::Nodes::Builder
...
+  # ```
+  def alias(anchor : String) : Nil
+    push_node Alias.new anchor
+  end
+
+  # Emits the scalar "<<" followed by an alias to the given *anchor*.
+  #
+  # See [YAML Merge](https://yaml.org/type/merge.html).
+  #
+  # ```crystal
```
Comment (Sija) [inline src/yaml/nodes/builder.cr]:
```suggestion
  # ```
```

File: src/yaml/builder.cr [hunk 13]
```diff
@@ -130,11 +130,45 @@ class YAML::Builder
     yield.tap { end_mapping }
   end
 
-  def alias(anchor : String)
+  # Emits an alias to the given *anchor*.
+  #
+  # ```crystal
```
Comment (Sija) [inline src/yaml/builder.cr]:
```suggestion
  # ```
```

File: src/yaml/builder.cr [hunk 14]
```diff
@@ -130,11 +130,45 @@ class YAML::Builder
...
+  # end
+  #
+  # yaml # => "---\nkey: *example\n"
+  # ```
+  def alias(anchor : String) : Nil
     LibYAML.yaml_alias_event_initialize(pointerof(@event), anchor)
     yaml_emit("alias")
   end
 
+  # Emits the scalar "<<" followed by an alias to the given *anchor*.
```
Comment (Sija) [inline src/yaml/builder.cr]:
```suggestion
  # Emits the scalar `"<<"` followed by an alias to the given *anchor*.
```

File: src/yaml/nodes/builder.cr [hunk 15]
```diff
@@ -22,6 +22,52 @@ class YAML::Nodes::Builder
...
+  #   nodes_builder.document.to_yaml builder
+  # end
+  #
+  # yaml # => "---\nkey: *foo\n"
+  # ```
+  def alias(anchor : String) : Nil
+    push_node Alias.new anchor
+  end
+
+  # Emits the scalar "<<" followed by an alias to the given *anchor*.
```
Comment (Sija) [inline src/yaml/nodes/builder.cr]:
```suggestion
  # Emits the scalar `"<<"` followed by an alias to the given *anchor*.
```

File: spec/std/yaml/nodes/builder_spec.cr (hunk 5 above)
Comment (Blacksmoke16) [reply spec/std/yaml/nodes/builder_spec.cr]: These would cause it to not compile since `alias` is a keyword

File: src/yaml/builder.cr (hunk 6 above)
Comment (Blacksmoke16) [reply src/yaml/builder.cr]: Used self just so its consistent with the alias call below.

Comment (Sija) [reply src/yaml/builder.cr]: It's redundant in both places, so why not omit it?

Comment (Blacksmoke16) [reply src/yaml/builder.cr]: 🤷‍♂ Consistency sake?

Comment (Sija) [reply src/yaml/builder.cr]: > EDIT: It's required for the `alias` method call for the same reason it needs `itself`.

Calling methods without a receiver implicitly uses `self`, so what would be the reason to do so?

Comment (Blacksmoke16) [reply src/yaml/builder.cr]: image

Comment (Blacksmoke16) [reply src/yaml/builder.cr]: image

Comment (Sija) [reply src/yaml/builder.cr]: My bad then, forgot it's a language keyword 🤦‍♂

**Pull request merged**